bpf/lib/lxc.h

```diff
@@ -65,13 +65,49 @@ static inline int is_valid_lxc_src_ipv4(struct iphdr *ip4)
 }
 #endif
 
+/**
+ * skb_redirect_to_proxy configures the skb with the proxy mark and proxy port
+ * number to ensure that the stack redirects the packet into the proxy.
+ *
+ * It is called from both ingress and egress side of endpoint devices.
+ *
+ * In regular veth mode:
+ * * To apply egress policy, the egressing endpoint configures the mark here,
+ *   which returns TC_ACT_OK to pass the packet to the stack in the context
+ *   of the source device (stack ingress).
+ * * To apply ingress policy, the egressing endpoint or netdev program tail
+ *   calls into the policy program which configures the mark here, which
+ *   returns TC_ACT_OK to pass the packet to the stack in the context of the
+ *   source device (netdev or egress endpoint device, stack ingress).
+ *
+ * In chaining mode with bridged endpoint devices:
+ * * To apply egress policy, the egressing endpoint configures the mark here,
+ *   which is propagated to skb->cb[] in the caller. The redirect() call here
+ *   redirects the packet to the ingress TC filter configured on the bridge
+ *   master device.
+ * * To apply ingress policy, the stack transmits the packet into the bridge
+ *   master device which tail calls into the policy program for the ingress
+ *   endpoint, which configures mark and cb[] as described for the egress path.
+ *   The redirect() call here redirects the packet to the ingress TC filter
+ *   configured on the bridge master device.
+ * * In both cases for bridged endpoint devices, the bridge master device has
+ *   a BPF program configured upon ingress to transfer the cb[] to the mark
+ *   before passing the traffic up to the stack towards the proxy.
+ */
 static inline int __inline__
 skb_redirect_to_proxy(struct __sk_buff *skb, __be16 proxy_port)
 {
 	skb->mark = MARK_MAGIC_TO_PROXY | proxy_port << 16;
+
+	cilium_dbg_capture(skb, DBG_CAPTURE_PROXY_PRE, proxy_port);
+
+#ifdef HOST_REDIRECT_TO_INGRESS
+	return redirect(HOST_IFINDEX, BPF_F_INGRESS);
+#else
+	cilium_dbg_capture(skb, DBG_CAPTURE_PROXY_POST, proxy_port);
```

Comment:
Are the PROXY_PRE and PROXY_POST debug capture intentionally back-to-back in this case?

Reply:
Yeah, arguably we could remove this but it means that we have matching pre/post messages instead of only a pre message in the regular veth case. In the flannel case, the post happens in context of another device which is handy for determining that the packet successfully reached the other logic.

Reply:
One option to shuffle it around a little better would be to only do the post message in the veth case here, and only do the pre message in the flannel case here. Then we'd cut down one debug message.